prov/usnic/src/usdf_wait.c

```diff
@@ -0,0 +1,338 @@
+ * on the domain because we have a context associated with the domain. At
+ * fi_wait_open time, we only have access to the fabric. It isn't guaranteed
+ * that a domain has been opened yet. The epoll context approach allows us to
+ * defer creating the completion channel for the CQ until CQ open time.
+ */
+int usdf_wait_open(struct fid_fabric *fabric, struct fi_wait_attr *attr,
+		struct fid_wait **waitset)
+{
+	struct usdf_wait *wait_priv;
+	struct usdf_fabric *fabric_priv;
+	int epfd;
+	int ret;
+
+	USDF_TRACE_SYS(FABRIC, "\n");
+
+	switch (attr->wait_obj) {
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		break;
+	default:
+		USDF_WARN_SYS(FABRIC, "unsupported wait object type\n");
+		ret = -FI_EINVAL;
+		goto error;
+	}
+
+	fabric_priv = fab_fidtou(fabric);
+	ret = usdf_wait_check_support(fabric_priv);
+	if (ret)
+		goto error;
+
+	epfd = epoll_create1(0);
+	if (epfd < 0) {
+		USDF_WARN_SYS(FABRIC, "failed to create epoll fd[%d]\n", errno);
+		ret = -errno;
+		goto error;
+	}
+
+	USDF_DBG_SYS(FABRIC, "successfully created epoll fd: %d\n", epfd);
+
+	wait_priv = calloc(1, sizeof(*wait_priv));
+	if (!wait_priv) {
+		USDF_WARN_SYS(FABRIC,
+				"unable to allocate memory for usdf_wait obj");
+		ret = -FI_ENOMEM;
+		goto calloc_fail;
+	}
+
+	wait_priv->wait_fid.fid.fclass = FI_CLASS_WAIT;
+	wait_priv->wait_fid.fid.ops = &usdf_wait_fi_ops;
+	wait_priv->wait_fid.ops = &usdf_wait_ops;
+	wait_priv->wait_fid.fid.context = 0;
+	wait_priv->wait_fabric = fabric_priv;
+	wait_priv->wait_obj = attr->wait_obj;
+	wait_priv->object.epfd = epfd;
+
+	atomic_initialize(&wait_priv->wait_refcnt, 0);
+	fastlock_init(&wait_priv->lock);
+	dlist_init(&wait_priv->list);
+
+	atomic_inc(&wait_priv->wait_fabric->fab_refcnt);
+
+	*waitset = &wait_priv->wait_fid;
+
+	return FI_SUCCESS;
+
+calloc_fail:
+	close(epfd);
+error:
+	*waitset = NULL;
+	return ret;
+}
+
+/* Close a wait object. Make sure all resources associated with the wait object
+ * have been closed.
+ */
+static int usdf_wait_close(struct fid *waitset)
+{
+	struct usdf_wait *wait_priv;
+
+	USDF_TRACE_SYS(FABRIC, "\n");
+	if (!waitset) {
+		USDF_WARN_SYS(FABRIC, "invalid input.\n");
+		return -FI_EINVAL;
+	}
+
+	wait_priv = wait_ftou(waitset);
+
+	if (atomic_get(&wait_priv->wait_refcnt) > 0) {
+		USDF_DBG_SYS(FABRIC,
+				"failed to close waitset with non-zero refcnt");
+		return -FI_EBUSY;
+	}
+
+	switch (wait_priv->wait_obj) {
+	case FI_WAIT_UNSPEC:
+	case FI_WAIT_FD:
+		close(wait_priv->object.epfd);
+		break;
+	default:
+		USDF_WARN_SYS(FABRIC,
+				"unsupported wait object type\n");
+		return -FI_EINVAL;
+	}
+
+	atomic_dec(&wait_priv->wait_fabric->fab_refcnt);
+	free(wait_priv);
+
+	return FI_SUCCESS;
+}
+
+static int usdf_wait_wait(struct fid_wait *fwait, int timeout)
+{
+	struct usdf_wait *wait;
+	struct epoll_event event;
+	int ret = FI_SUCCESS;
+	int nevents;
+
+	USDF_TRACE_SYS(FABRIC, "\n");
+	wait = wait_ftou(fwait);
+
+	ret = usdf_wait_trywait(&fwait->fid);
```

Comment:
> Do you need a check before this to ensure it's a supported type of waitset?

Looked at this one more in depth and I don't need to check it here. The type of item being passed in is `fid_wait` meaning it's a waitset and the `usdf_wait_trywait` will go through each fid in the waitset and check all of them. If any of them are invalid, then it will error with `-FI_EINVAL` which this function will propogate up.